### PR TITLE
fix: Fix `__hash__` methods

### LINKED ISSUE #2506: __hash__ methods broken
## Expected Behavior

The `__hash__` methods should not be implemented like this:
```
def __hash__(self) -> int:
    return hash((id(self), self.name))
```

Objects with the `__hash__` method implemented in such a way are not being deduplicated correctly in e.g. sets and dicts.

### Specifications

- Version:
- Platform:
- Subsystem:

### PR DESCRIPTION
**Pull request by felixwang9817**
**What this PR does / why we need it**: This PR fixes the broken `__hash__` methods in all Feast objects.

**Which issue(s) this PR fixes**:

Fixes #2506

### CODE REVIEW

**Review by achals — APPROVED** [review]
/lgtm

**Review by achals — APPROVED** [review]
/lgtm

File: sdk/python/feast/saved_dataset.py [hunk 1]
```diff
@@ -92,17 +92,23 @@ def __str__(self):
         return str(MessageToJson(self.to_proto()))
 
     def __hash__(self):
-        return hash((id(self), self.name))
+        return hash((self.name))
```
Comment (judahrand) [inline sdk/python/feast/saved_dataset.py:95]: Unnecessary brackets?

**Pull request merged**